iceoryx2-ffi/cxx/include/iox/slice.hpp

```diff
@@ -16,38 +16,142 @@
 #include "iox/assertions_addendum.hpp"
 
 #include <cstdint>
+#include <type_traits>
 
 namespace iox {
+
+/// @brief A class representing a slice of contiguous elements of type T.
+///
+/// A Slice provides a view into a contiguous sequence of elements without owning the memory.
+/// It allows for efficient access and iteration over a portion of a contiguous data structure.
+///
+/// @tparam T The type of elements in the slice. Can be const-qualified for read-only slices.
 template <typename T>
 class Slice {
   public:
-    using Iterator = T*;
-    using ConstIterator = const T*;
-    using ValueType = T;
-
-    auto size() const -> uint64_t {
-        IOX_TODO();
-    }
-    auto operator[](const uint64_t n) const -> const T& {
-        IOX_TODO();
-    }
-    auto operator[](const uint64_t n) -> T& {
-        IOX_TODO();
-    }
-    auto begin() -> Iterator {
-        IOX_TODO();
-    }
-    auto begin() const -> ConstIterator {
-        IOX_TODO();
-    }
-    auto end() -> Iterator {
-        IOX_TODO();
-    }
-    auto end() const -> ConstIterator {
-        IOX_TODO();
-    }
+    using Iterator = std::conditional_t<std::is_const_v<T>, const T*, T*>;
+    using ValueType = std::remove_const_t<T>;
+
+    /// @brief Constructs a Slice object.
+    /// @param[in] data Pointer to the beginning of the data.
+    /// @param[in] number_of_elements The number of elements in the slice.
+    Slice(T* data, uint64_t number_of_elements);
+
+    /// @brief Returns the total number of bytes occupied by the slice.
+    /// @return The number of bytes occupied by the slice, rounded up to the nearest alignment boundary.
+    auto number_of_bytes() const -> uint64_t;
+
+    /// @brief Returns the number of elements in the slice.
+    /// @return The number of elements in the slice.
+    auto number_of_elements() const -> uint64_t;
+
+    /// @brief Accesses the element at the specified index (const version).
+    /// @param[in] n The index of the element to access.
+    /// @return A const reference to the element at the specified index.
+    /// @pre The index must be less than the number of elements in the slice.
+    auto operator[](uint64_t n) const -> std::conditional_t<std::is_const_v<T>, const ValueType&, ValueType&>;
+
+    /// @brief Accesses the element at the specified index (non-const version).
+    /// @param[in] n The index of the element to access.
+    /// @return A reference to the element at the specified index.
+    /// @pre The index must be less than the number of elements in the slice.
+    auto operator[](uint64_t n) -> std::conditional_t<std::is_const_v<T>, const ValueType&, ValueType&>;
+
+    /// @brief Returns an iterator to the beginning of the slice (const version).
+    /// @return An iterator pointing to the first element of the slice.
+    auto begin() const -> Iterator;
```

Comment:
If the type argument is a non-const `T` then this would again violate the API contract and a user could modify the const slice by accident just by iterating over it. This can easily be fixed by using here
```cxx
auto begin() const -> ConstIterator;
```